### PR TITLE
Implement an EGL resource context for the Linux shell.

### PR DESCRIPTION
**Pull request by robert-ancell**
Fixes https://github.com/flutter/flutter/issues/54855

### CODE REVIEW

**Comment by robert-ancell** [comment]
I'm no expert on EGL, but I've attempted to copy the implementations of the shared context in the other shells. Everything looks the same as before, so, job done?

File: shell/platform/linux/fl_renderer.cc [hunk 1]
```diff
@@ -52,29 +94,56 @@ static gboolean fl_renderer_real_start(FlRenderer* self, GError** error) {
...
                   EGL_CONTEXT_CLIENT_VERSION, &value);
 
+  const EGLint shared_context_attribs[] = {EGL_WIDTH, 1, EGL_HEIGHT, 1,
+                                           EGL_NONE};
+  priv->resource_surface = eglCreatePbufferSurface(
+      priv->egl_display, egl_config, shared_context_attribs);
+  if (priv->resource_surface == nullptr) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to create EGL resource surface: %s", get_egl_error());
+    return FALSE;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_renderer.cc:137]: Resource context failures don't need to be fatal; we can log the error and then proceed without the resource context.

File: shell/platform/linux/fl_renderer.cc [hunk 2]
```diff
@@ -99,7 +168,21 @@ gboolean fl_renderer_make_current(FlRenderer* self, GError** error) {
...
+                "Failed to make EGL context current: %s", get_egl_error());
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+gboolean fl_renderer_make_resource_current(FlRenderer* self, GError** error) {
+  FlRendererPrivate* priv =
+      static_cast<FlRendererPrivate*>(fl_renderer_get_instance_private(self));
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_renderer.cc:180]: And then early return here if it's null.

**Comment by stuartmorgan** [comment]
Please add the issue reference to the PR.

File: shell/platform/linux/fl_renderer.cc (hunk 2 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_renderer.cc:179]: Not sure what this comment is referring to?

Comment (stuartmorgan) [reply shell/platform/linux/fl_renderer.cc:179]: If `resource_surface` wasn't successfully created, none of the code below should be executed.

File: shell/platform/linux/fl_renderer.cc [hunk 3]
```diff
@@ -52,29 +94,55 @@ static gboolean fl_renderer_real_start(FlRenderer* self, GError** error) {
...
+  if (priv->resource_surface == nullptr) {
+    g_warning(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+              "Failed to create EGL resource surface: %s", get_egl_error());
+  }
+  priv->resource_context = eglCreateContext(
+      priv->egl_display, egl_config, priv->egl_context, context_attributes);
+  if (priv->resource_context == nullptr) {
+    g_warning(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+              "Failed to create EGL resource context: %s", get_egl_error());
+    return FALSE;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_renderer.cc:143]: This shouldn't return false.

File: shell/platform/linux/fl_renderer.cc [hunk 4]
```diff
@@ -52,29 +94,55 @@ static gboolean fl_renderer_real_start(FlRenderer* self, GError** error) {
...
 
+  const EGLint shared_context_attribs[] = {EGL_WIDTH, 1, EGL_HEIGHT, 1,
+                                           EGL_NONE};
+  priv->resource_surface = eglCreatePbufferSurface(
+      priv->egl_display, egl_config, shared_context_attribs);
+  if (priv->resource_surface == nullptr) {
+    g_warning(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+              "Failed to create EGL resource surface: %s", get_egl_error());
+  }
+  priv->resource_context = eglCreateContext(
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_renderer.cc:138]: Presumably all of this block should only run if `resource_surface` is non-null.

File: shell/platform/linux/fl_renderer.cc (hunk 2 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_renderer.cc:179]: Ah, GitHub is showing this on a different line but I know what you're referring to.

**Pull request merged**